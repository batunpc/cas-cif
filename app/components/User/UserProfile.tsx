import React from "react";
import { graphql, useFragment } from "react-relay";
import type { UserProfile_user$key } from "__generated__/UserProfile_user.graphql";
import Button from "@button-inc/bcgov-theme/Button";
import Grid from "@button-inc/bcgov-theme/Grid";

interface Props {
  user: UserProfile_user$key;
}

const UserProfile: React.FC<Props> = ({ user }) => {
  const { firstName, lastName, emailAddress } =
    useFragment(
      graphql`
        fragment UserProfile_user on CifUser {
          firstName
          lastName
          emailAddress
        }
      `,
      user
    ) || {};

  if (user === null) return null;

  return (
    <>
      <Grid className="name-display" justify="end">
        <Grid.Row>
          {lastName}, {firstName}
        </Grid.Row>
        <Grid.Row>{emailAddress}</Grid.Row>
      </Grid>
      <form action="/logout" method="post">
        <Button variant="secondary-inverse" type="submit">
          Logout
        </Button>
      </form>
      <style jsx>{`
        :global(.name-display) {
          margin-right: 20px;
        }
        form {
          margin-top: auto;
          margin-bottom: auto;
        }
      `}</style>
    </>
  );
};

export default UserProfile;
