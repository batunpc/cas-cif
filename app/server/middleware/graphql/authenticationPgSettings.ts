import { getPriorityGroup } from "../../../lib/userGroups";
import { getUserGroups } from "../../helpers/userGroupAuthentication";
import groupData from "../../../data/groups.json";
import { isAuthenticated } from "@bcgov-cas/sso-express";
import type { Request } from "express";

const AUTH_BYPASS_COOKIE = "mocks.auth";

const AS_CIF_INTERNAL = process.argv.includes("AS_CIF_INTERNAL");

const AS_CIF_EXTERNAL = process.argv.includes("AS_CIF_EXTERNAL");

const AS_CIF_ADMIN = process.argv.includes("AS_CIF_ADMIN");

const AS_UNAUTHORIZED_IDIR = process.argv.includes("AS_UNAUTHORIZED_IDIR");
const AS_CYPRESS = process.argv.includes("AS_CYPRESS");

const allowCypressForRole = (roleName: string, req: Request) => {
  return AS_CYPRESS && req.cookies[AUTH_BYPASS_COOKIE] === roleName;
};

const authenticationPgSettings = (req: Request) => {
  if (AS_CIF_INTERNAL || allowCypressForRole("cif_internal", req)) {
    return {
      "jwt.claims.sub": "00000000-0000-0000-0000-000000000000",
      "jwt.claims.user_groups": "cif_internal",
      "jwt.claims.priority_group": "cif_internal",
      role: "cif_internal",
    };
  }

  if (AS_CIF_EXTERNAL || allowCypressForRole("cif_external", req)) {
    return {
      "jwt.claims.sub": "00000000-0000-0000-0000-000000000001",
      "jwt.claims.user_groups": "cif_external",
      "jwt.claims.priority_group": "cif_external",
      role: "cif_external",
    };
  }

  if (AS_CIF_ADMIN || allowCypressForRole("cif_admin", req)) {
    return {
      "jwt.claims.sub": "00000000-0000-0000-0000-000000000002",
      "jwt.claims.user_groups": "cif_admin",
      "jwt.claims.priority_group": "cif_admin",
      role: "cif_admin",
    };
  }

  if (AS_UNAUTHORIZED_IDIR || allowCypressForRole("unauthorized_idir", req)) {
    return {
      "jwt.claims.sub": "00000000-0000-0000-0000-000000000000",
      "jwt.claims.user_groups": "UNAUTHORIZED_IDIR",
      "jwt.claims.priority_group": "UNAUTHORIZED_IDIR",
      role: "ciip_guest",
    };
  }

  const groups = getUserGroups(req);
  const priorityGroup = getPriorityGroup(groups);

  const claimsSettings = {
    role: groupData[priorityGroup].pgRole,
  };
  if (!isAuthenticated(req))
    return {
      ...claimsSettings,
    };

  const claims = req.claims;

  claims.user_groups = groups.join(",");
  claims.priority_group = priorityGroup;

  const properties = [
    "jti",
    "exp",
    "nbf",
    "iat",
    "iss",
    "aud",
    "sub",
    "typ",
    "azp",
    "auth_time",
    "session_state",
    "acr",
    "email_verified",
    "name",
    "preferred_username",
    "given_name",
    "family_name",
    "email",
    "broker_session_id",
    "user_groups",
    "priority_group",
  ];
  properties.forEach((property) => {
    claimsSettings[`jwt.claims.${property}`] = claims[property];
  });

  return {
    ...claimsSettings,
  };
};

export default authenticationPgSettings;
