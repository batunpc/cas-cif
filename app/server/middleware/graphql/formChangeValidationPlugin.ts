import { makeWrapResolversPlugin } from "postgraphile";
import validationSchemas from "../../../data/jsonSchemaForm/validationSchemas";
import Ajv, { ErrorObject } from "ajv";

const ajv = new Ajv({ allErrors: true });

function validateRecord(schemaName: string, formData: any): ErrorObject[] {
  const schema = validationSchemas[schemaName];

  // ajv caches compiled schemas on first instantiation, we don't need to
  // precompile schemas in advance
  const validate = ajv.compile(schema);
  const valid = validate(formData);

  return valid ? [] : validate.errors;
}

export const FormChangeValidationPlugin = makeWrapResolversPlugin(
  (context) => {
    if (
      context.scope.isRootMutation &&
      context.scope.fieldName === "updateFormChange"
    ) {
      return {
        // There is no need to pass a context to the resolver at this time
      };
    }
    return null;
  },
  () => async (resolver: any, source, args, context, resolveInfo) => {
    // If the mutation doesn't change the form data, we don't need to re-validate
    if (args.input?.formChangePatch?.newFormData === undefined)
      return resolver();

    const {
      identifiers: [formChangeId],
    } = (resolveInfo as any).graphile.build.getTypeAndIdentifiersFromNodeId(
      args.input.id
    );

    // This is the recommended way to fetch data ahead of running the
    // resolver of the GraphQL mutation - see first "IMPORTANT" note here:
    // https://www.graphile.org/postgraphile/make-extend-schema-plugin/#the-selectgraphqlresultfromtable-helper
    const { pgClient } = context;
    const {
      rows: [formChangeRecord],
    } = await pgClient.query(
      `select json_schema_name from cif.form_change where id = $1`,
      [formChangeId]
    );

    const errors = validateRecord(
      formChangeRecord.json_schema_name,
      args.input.formChangePatch.newFormData
    );

    args.input.formChangePatch = {
      ...args.input.formChangePatch,
      validationErrors: errors,
    };

    const result = await resolver(source, args, context, resolveInfo);
    return result;
  }
);
