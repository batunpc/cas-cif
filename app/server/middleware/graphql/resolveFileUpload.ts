import config from "../../../config";

async function saveRemoteFile({ stream }) {
  const response = await fetch(
    `${config.get("storageApiHost")}/api/v1/attachments/upload`,
    {
      method: "POST",
      headers: {
        "api-key": config.get("storageApiKey"),
        "Content-Type": "multipart/form-data",
      },
      body: stream,
    }
  );
  try {
    return await response.json();
  } catch (e) {
    console.error(e);
  }
}

export default async function resolveFileUpload(upload) {
  const { createReadStream } = upload;
  const stream = createReadStream();

  // Save tile to remote storage system
  const { uuid } = await saveRemoteFile({ stream });

  return uuid;
}
