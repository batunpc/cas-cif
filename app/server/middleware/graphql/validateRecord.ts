import Ajv, { ErrorObject } from "ajv";
import { KEY_EMAIL, KEY_PHONE_NO, KEY_RFP_DIGITS, REGEX_EMAIL, REGEX_PHONE_NO, REGEX_RFP_DIGITS } from "data/validation-constants";

const ajv = new Ajv({ allErrors: true });
// AJV needs to be made aware of any custom formats used in the schema
ajv.addFormat(KEY_RFP_DIGITS, REGEX_RFP_DIGITS);
ajv.addFormat(KEY_EMAIL, REGEX_EMAIL);
ajv.addFormat(KEY_PHONE_NO, REGEX_PHONE_NO);

const validateRecord: (schema: any, formData: any) => ErrorObject[] = (
  schema,
  formData
) => {
  // ajv caches compiled schemas on first instantiation, we don't need to
  // precompile schemas in advance
  const validate = ajv.compile(schema);
  const valid = validate(formData);

  return valid ? [] : validate.errors;
};

export default validateRecord;
