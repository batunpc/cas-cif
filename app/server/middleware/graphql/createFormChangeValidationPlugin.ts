import { GraphQLObjectType, GraphQLResolveInfo } from "graphql";
import { Context, makeWrapResolversPlugin } from "postgraphile";
import validateRecord from "./validateRecord";
import databaseSchemaConfiguration from "./databaseSchemaConfiguration";
import validationSchemas from "../../../data/jsonSchemaForm/validationSchemas";

const DATABASE_SCHEMAS = databaseSchemaConfiguration;

export const filter = (context: Context<GraphQLObjectType<any, any>>) => {
  if (
    context.scope.isRootMutation &&
    context.scope.fieldName === "createFormChange"
  ) {
    return {
      // There is no need to pass a context to the resolver at this time
    };
  }
  return null;
};

export const resolverWrapperGenerator =
  () =>
  async (
    resolver: any,
    source,
    args: { [argName: string]: any },
    context,
    resolveInfo: GraphQLResolveInfo
  ) => {
    if (!args.input.jsonSchemaName) {
      throw new Error("JSON schema must be set in the mutation input");
    }
    const { pgClient } = context;

    let jsonSchema;

    if (DATABASE_SCHEMAS.includes(args.input.jsonSchemaName)) {
      const {
        rows: [fetchedRecord],
      } = await pgClient.query(
        `select json_schema from cif.form
         where slug = $1`,
        [args.input.jsonSchemaName]
      );
      jsonSchema = fetchedRecord.json_schema.schema;
    } else jsonSchema = validationSchemas[args.input.jsonSchemaName];

    const errors = validateRecord(jsonSchema, args.input.newFormData ?? {});

    args.input.validationErrors = errors;

    const result = await resolver(source, args, context, resolveInfo);
    return result;
  };

const CreateFormChangeValidationPlugin = makeWrapResolversPlugin(
  filter,
  resolverWrapperGenerator
);

export default CreateFormChangeValidationPlugin;
