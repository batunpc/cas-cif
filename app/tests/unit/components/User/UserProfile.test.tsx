import { render } from "@testing-library/react";
import UserProfile from "components/User/UserProfile";
import { mockRandom } from "jest-mock-random";
import { UserProfile_user } from "__generated__/UserProfile_user.graphql";

beforeEach(() => {
  mockRandom([0.3, 0.1, 0.4, 0.1, 0.5, 0.9]);
});

describe("The UserProfile component", () => {
  it("matches the snapshot", () => {
    const userData: UserProfile_user = {
      emailAddress: "a@a.a",
      firstName: "a",
      lastName: "a",
      " $refType": "UserProfile_user",
    };

    jest
      .spyOn(require("react-relay"), "useFragment")
      .mockImplementation(() => userData);

    const componentUnderTest = render(<UserProfile user={null} />);

    expect(componentUnderTest.container).toMatchSnapshot();
  });
});
