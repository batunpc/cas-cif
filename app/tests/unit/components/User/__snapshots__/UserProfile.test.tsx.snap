// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The UserProfile component matches the snapshot 1`] = `
<div>
  <div
    class="sc-iBPTik cirWLn pg-grid-container name-display"
  >
    <div
      class="sc-fubCzh sc-pGacB kCFBWZ elfPxV pg-grid-row"
    >
      a
      , 
      a
    </div>
    <div
      class="sc-fubCzh sc-pGacB kCFBWZ elfPxV pg-grid-row"
    >
      a@a.a
    </div>
  </div>
  <form
    action="/logout"
    class="jsx-4065309065"
    method="post"
  >
    <button
      class="sc-gsTEea gemKec pg-button"
      id="eeeeeeeeeeg3lllllllllm"
      type="submit"
    >
      Logout
    </button>
  </form>
</div>
`;
